controller/gke-cluster-config-handler.go

```diff
@@ -0,0 +1,349 @@
+				// the update is valid in that the controller should retry but there is
+				// no actionable resolution as far as a user is concerned. An update
+				// that has either been initiated by gke-operator or another source is
+				// already in progress. It is possible an update is not being immediately
+				// reflected in upstream cluster state. The config object will reenter the
+				// controller and then the controller will wait for the update to finish.
+				message = err.Error()
+			}
+		}
+
+		if config.Status.FailureMessage == message {
+			return config, err
+		}
+
+		if message != "" {
+			config = config.DeepCopy()
+			if config.Status.Phase == gkeConfigActivePhase {
+				// can assume an update is failing
+				config.Status.Phase = gkeConfigUpdatingPhase
+			}
+		}
+		config.Status.FailureMessage = message
+
+		var recordErr error
+		config, recordErr = h.gkeCC.UpdateStatus(config)
+		if recordErr != nil {
+			logrus.Errorf("Error recording gkecc [%s] failure message: %s", config.Name, recordErr.Error())
+		}
+		return config, err
+	}
+}
+
+func (h *Handler) OnGkeConfigRemoved(key string, config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	if config.Spec.Imported {
+		logrus.Infof("cluster [%s] is imported, will not delete GKE cluster", config.Name)
+		return config, nil
+	}
+	if config.Status.Phase == gkeConfigNotCreatedPhase {
+		// The most likely context here is that the cluster already existed in GKE, so we shouldn't delete it
+		logrus.Warnf("cluster [%s] never advanced to creating status, will not delete GKE cluster", config.Name)
+		return config, nil
+	}
+
+	logrus.Infof("deleting cluster [%s]", config.Name)
+
+	//TODO: delete cluster
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := getServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	logrus.Debugf("Removing cluster %v from project %v, region/zone %v", config.Spec.ClusterName, config.Spec.ProjectID, config.Spec.Region)
+	operation, err := waitClusterRemoveExp(ctx, svc, config)
+
+	if err != nil && !strings.Contains(err.Error(), "notFound") {
+		return config, err
+	} else if err == nil {
+		logrus.Debugf("Cluster %v delete is called. Status Code %v", config.Spec.ClusterName, operation.HTTPStatusCode)
+	} else {
+		logrus.Debugf("Cluster %s doesn't exist", config.Spec.ClusterName)
+	}
+
+	return config, nil
+}
+
+func (h *Handler) create(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := getServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	operation, err := svc.Projects.Locations.Clusters.Create(
+		locationRRN(config.Spec.ProjectID, config.Spec.Region),
+		generateClusterCreateRequest(config)).Context(ctx).Do()
+
+	if err != nil && !strings.Contains(err.Error(), "alreadyExists") {
```

Comment:
I guess I don't understand. this statement would only get hit if you created the CRD and the cluster already exists. it would not be hit if you created one in GKE and did nothing in rancher.